// @datadog/browser-rum-core/cjs/domain/resource/retrieveInitialDocumentResourceTiming.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.retrieveInitialDocumentResourceTiming = retrieveInitialDocumentResourceTiming;
const browser_core_1 = require("@datadog/browser-core");
const performanceObservable_1 = require("../../browser/performanceObservable");
const getDocumentTraceId_1 = require("../tracing/getDocumentTraceId");
const performanceUtils_1 = require("../../browser/performanceUtils");
const resourceUtils_1 = require("./resourceUtils");
function retrieveInitialDocumentResourceTiming(configuration, callback, getNavigationEntryImpl = performanceUtils_1.getNavigationEntry) {
    (0, browser_core_1.runOnReadyState)(configuration, 'interactive', () => {
        const navigationEntry = getNavigationEntryImpl();
        const entry = Object.assign(navigationEntry.toJSON(), {
            entryType: performanceObservable_1.RumPerformanceEntryType.RESOURCE,
            initiatorType: resourceUtils_1.FAKE_INITIAL_DOCUMENT,
            // The ResourceTiming duration entry should be `responseEnd - startTime`. With
            // NavigationTiming entries, `startTime` is always 0, so set it to `responseEnd`.
            duration: navigationEntry.responseEnd,
            traceId: (0, getDocumentTraceId_1.getDocumentTraceId)(document),
            toJSON: () => ({ ...entry, toJSON: undefined }),
        });
        callback(entry);
    });
}
